control-plane/csi-driver/src/bin/node/fsfreeze/bin/ioctl.rs

```diff
@@ -0,0 +1,61 @@
+use crate::fsfreeze::bin::FsFreezeOpt;
+use nix::{errno::Errno, ioctl_readwrite};
+use std::{fs, os::fd::AsRawFd, path::PathBuf};
+use tracing::error;
+
+// Nix macro to issue FIFREEZE ioctl.
+ioctl_readwrite!(fs_freeze, 'X', 119, i32);
+// Nix macro to issue FITHAW ioctl.
+ioctl_readwrite!(fs_thaw, 'X', 120, i32);
+
+/// Preflight checks to ensure that fsfreeze call would be made in a desirable condition.
+pub fn fsfreeze_preflight_check(volume_id: &str, nqn: &str) -> Result<(), Errno> {
+    let subsystems =
+        nvmeadm::nvmf_subsystem::Subsystem::try_from_nqn(nqn).map_err(|_| Errno::ENODATA)?;
+    if subsystems.iter().any(|subsystem| subsystem.state == "live") {
+        return Ok(());
+    }
+    error!(
+        volume_id = volume_id,
+        nqn = nqn,
+        "Preflight check for fsfreeze failed, subsystem has no live path"
+    );
+    Err(Errno::EHOSTUNREACH)
+}
+
+/// Issue FIFREEZE and FITHAW ioctl on the mount_path based on the command.
+pub async fn fsfreeze_ioctl(command: FsFreezeOpt, mount_path: PathBuf) -> Result<(), Errno> {
+    let blocking_task = tokio::task::spawn_blocking(move || {
+        let file = fs::File::open(&mount_path).map_err(|error| {
+            error!(
+                "Failed to get file descriptor path: {}, {}",
+                mount_path.display(),
+                error.to_string()
+            );
+            Errno::EBADFD
+        })?;
+        match command {
+            FsFreezeOpt::Freeze => match unsafe { fs_freeze(file.as_raw_fd(), &mut 0) } {
+                Ok(_) => Ok(()),
+                Err(Errno::EBUSY) => Ok(()),
+                Err(errno) => Err(errno),
+            },
+            FsFreezeOpt::Unfreeze => loop {
+                match unsafe { fs_thaw(file.as_raw_fd(), &mut 0) } {
+                    Ok(val) if val != 0 => return Ok(()),
+                    Err(Errno::EINVAL) => return Ok(()),
+                    Err(errno) => return Err(errno),
+                    _ => {}
```

Comment:
nit: might be clearer here
```suggestion
                    Ok(val) => { if val != 0 return Ok(()); },
                    Err(Errno::EINVAL) => return Ok(()),
                    Err(errno) => return Err(errno),
```
or 
```suggestion
                    Ok(val) if val != 0 => return Ok(()),
                    Ok(val) => {},
                    Err(Errno::EINVAL) => return Ok(()),
                    Err(errno) => return Err(errno),
```